File: server/lib/scope.js
/* ============================================================================
   Scope parsing and enforcement.

   Mirrors CT.net on the client, but this copy is the one that matters: the
   browser can be bypassed, this cannot.
   ========================================================================= */
'use strict';

const MAX_PREFIX = 16;        // nothing broader than a /16
const MAX_ADDRESSES = 65536;

/* Ranges an assessment may touch without an explicit override. */
const SAFE_RANGES = [
  ['10.0.0.0', 8, 'RFC 1918 private'],
  ['172.16.0.0', 12, 'RFC 1918 private'],
  ['192.168.0.0', 16, 'RFC 1918 private'],
  ['100.64.0.0', 10, 'RFC 6598 carrier-grade NAT'],
  ['169.254.0.0', 16, 'RFC 3927 link-local'],
  ['127.0.0.0', 8, 'loopback']
];

function ipToInt(ip) {
  const parts = String(ip).trim().split('.');
  if (parts.length !== 4) throw new Error('Not an IPv4 address: ' + ip);
  let n = 0;
  for (const part of parts) {
    if (!/^\d{1,3}$/.test(part)) throw new Error('Not an IPv4 address: ' + ip);
    const octet = Number(part);
    if (octet > 255) throw new Error('Octet out of range: ' + ip);
    n = (n * 256) + octet;
  }
  return n >>> 0;
}

function intToIp(value) {
  const n = value >>> 0;
  return [(n >>> 24) & 255, (n >>> 16) & 255, (n >>> 8) & 255, n & 255].join('.');
}

function maskFor(prefix) {
  return prefix === 0 ? 0 : ((0xFFFFFFFF << (32 - prefix)) >>> 0);
}

function rangeOf(ip, prefix) {
  const mask = maskFor(prefix);
  const network = (ipToInt(ip) & mask) >>> 0;
  const broadcast = (network | (~mask >>> 0)) >>> 0;
  return { network, broadcast, mask };
}

/** Which safe range an address falls in, or null if it is outside all of them. */
function safeRangeOf(ip) {
  const value = ipToInt(ip);
  for (const [base, prefix, label] of SAFE_RANGES) {
    const { network, broadcast } = rangeOf(base, prefix);
    if (value >= network && value <= broadcast) return label;
  }
  return null;
}

function cidrInfo(text) {
  const raw = String(text).trim();
  const slash = raw.indexOf('/');
  const addr = slash === -1 ? raw : raw.slice(0, slash);
  const prefix = slash === -1 ? 32 : Number(raw.slice(slash + 1));

  if (!Number.isInteger(prefix) || prefix < 0 || prefix > 32) {
    throw new Error('Prefix must be between /0 and /32: ' + raw);
  }
  const { network, broadcast, mask } = rangeOf(addr, prefix);
  const total = broadcast - network + 1;

  return {
    input: raw,
    prefix,
    network: intToIp(network),
    broadcast: intToIp(broadcast),
    netmask: intToIp(mask),
    firstHost: intToIp(prefix >= 31 ? network : network + 1),
    lastHost: intToIp(prefix >= 31 ? broadcast : broadcast - 1),
    totalAddresses: total,
    usableHosts: prefix >= 31 ? total : Math.max(0, total - 2),
    safeRange: safeRangeOf(intToIp(network))
  };
}

/**
 * Parse an operator-supplied scope into the exact address list that may be
 * touched. Comma separated: CIDR blocks, single addresses, or a-b ranges.
 *
 * @param {string} text
 * @param {{allowPublic?: boolean}} [options]
 */
function parseScope(text, options) {
  const opts = options || {};
  const pieces = String(text || '').split(',').map((s) => s.trim()).filter(Boolean);
  if (!pieces.length) throw scopeError('No scope was supplied.');

  const entries = [];
  const seen = new Set();
  const addresses = [];

  for (const piece of pieces) {
    let first;
    let last;
    let info = null;

    const dash = piece.indexOf('-');
    if (dash !== -1 && piece.indexOf('/') === -1) {
      first = ipToInt(piece.slice(0, dash).trim());
      last = ipToInt(piece.slice(dash + 1).trim());
      if (last < first) throw scopeError('Range runs backwards: ' + piece);
    } else {
      info = cidrInfo(piece);
      if (info.prefix < MAX_PREFIX) {
        throw scopeError('Scope ' + piece + ' is broader than /' + MAX_PREFIX +
          '. Narrow it to the segment you are authorized to assess.');
      }
      first = ipToInt(info.network);
      last = ipToInt(info.broadcast);
      // A /31 or /32 addresses every value; anything wider excludes network
      // and broadcast, which are not hosts.
      if (info.prefix < 31) { first += 1; last -= 1; }
    }

    const count = last - first + 1;
    if (count <= 0) throw scopeError('Scope ' + piece + ' contains no usable address.');

    for (let value = first; value <= last; value++) {
      const ip = intToIp(value);
      if (!opts.allowPublic && !safeRangeOf(ip)) {
        throw scopeError('Address ' + ip + ' is outside private address space. ' +
          'Start the service with --allow-public to assess routable addresses, ' +
          'and only with written authorization for those systems.');
      }
      if (seen.has(ip)) continue;
      seen.add(ip);
      addresses.push(ip);
      if (addresses.length > MAX_ADDRESSES) {
        throw scopeError('Scope exceeds ' + MAX_ADDRESSES + ' addresses.');
      }
    }

    entries.push({ input: piece, info, first: intToIp(first), last: intToIp(last), count });
  }

  return {
    label: entries.map((e) => e.input).join(', '),
    entries,
    addresses,
    total: addresses.length,
    contains: (ip) => seen.has(ip)
  };
}

function scopeError(message) {
  const e = new Error(message);
  e.code = 'SCOPE';
  return e;
}

module.exports = {
  MAX_PREFIX, MAX_ADDRESSES, SAFE_RANGES,
  ipToInt, intToIp, maskFor, cidrInfo, safeRangeOf, parseScope
};

File: server/py/scope.py
"""Scope parsing and enforcement.

Mirrors the JavaScript service. This copy is the one that matters: the UI
can be bypassed by anything that reaches the port, this cannot.
"""

MAX_PREFIX = 16
MAX_ADDRESSES = 65536

# Ranges an assessment may touch without an explicit override.
SAFE_RANGES = [
    ("10.0.0.0", 8, "RFC 1918 private"),
    ("172.16.0.0", 12, "RFC 1918 private"),
    ("192.168.0.0", 16, "RFC 1918 private"),
    ("100.64.0.0", 10, "RFC 6598 carrier-grade NAT"),
    ("169.254.0.0", 16, "RFC 3927 link-local"),
    ("127.0.0.0", 8, "loopback"),
]


class ScopeError(ValueError):
    pass


def ip_to_int(ip):
    parts = str(ip).strip().split(".")
    if len(parts) != 4:
        raise ScopeError("Not an IPv4 address: %s" % ip)
    value = 0
    for part in parts:
        if not part.isdigit() or len(part) > 3:
            raise ScopeError("Not an IPv4 address: %s" % ip)
        octet = int(part)
        if octet > 255:
            raise ScopeError("Octet out of range: %s" % ip)
        value = value * 256 + octet
    return value


def int_to_ip(value):
    value &= 0xFFFFFFFF
    return "%d.%d.%d.%d" % (
        (value >> 24) & 255, (value >> 16) & 255, (value >> 8) & 255, value & 255)


def mask_for(prefix):
    return 0 if prefix == 0 else ((0xFFFFFFFF << (32 - prefix)) & 0xFFFFFFFF)


def safe_range_of(ip):
    """Which private range an address falls in, or None if it is routable."""
    value = ip_to_int(ip)
    for base, prefix, label in SAFE_RANGES:
        mask = mask_for(prefix)
        network = ip_to_int(base) & mask
        broadcast = network | (~mask & 0xFFFFFFFF)
        if network <= value <= broadcast:
            return label
    return None


def cidr_info(text):
    raw = str(text).strip()
    if "/" in raw:
        addr, _, prefix_text = raw.partition("/")
        if not prefix_text.isdigit():
            raise ScopeError("Prefix must be a number: %s" % raw)
        prefix = int(prefix_text)
    else:
        addr, prefix = raw, 32
    if not 0 <= prefix <= 32:
        raise ScopeError("Prefix must be between /0 and /32: %s" % raw)

    mask = mask_for(prefix)
    network = ip_to_int(addr) & mask
    broadcast = network | (~mask & 0xFFFFFFFF)
    total = broadcast - network + 1

    return {
        "input": raw,
        "prefix": prefix,
        "network": int_to_ip(network),
        "broadcast": int_to_ip(broadcast),
        "netmask": int_to_ip(mask),
        "firstHost": int_to_ip(network if prefix >= 31 else network + 1),
        "lastHost": int_to_ip(broadcast if prefix >= 31 else broadcast - 1),
        "totalAddresses": total,
        "usableHosts": total if prefix >= 31 else max(0, total - 2),
        "safeRange": safe_range_of(int_to_ip(network)),
    }


class Scope(object):
    def __init__(self, label, entries, addresses):
        self.label = label
        self.entries = entries
        self.addresses = addresses
        self.total = len(addresses)
        self._set = set(addresses)

    def contains(self, ip):
        return ip in self._set


def parse_scope(text, allow_public=False):
    """Parse an operator scope into the exact addresses that may be touched."""
    pieces = [p.strip() for p in str(text or "").split(",") if p.strip()]
    if not pieces:
        raise ScopeError("No scope was supplied.")

    entries, addresses, seen = [], [], set()

    for piece in pieces:
        info = None
        if "-" in piece and "/" not in piece:
            low, _, high = piece.partition("-")
            first, last = ip_to_int(low.strip()), ip_to_int(high.strip())
            if last < first:
                raise ScopeError("Range runs backwards: %s" % piece)
        else:
            info = cidr_info(piece)
            if info["prefix"] < MAX_PREFIX:
                raise ScopeError(
                    "Scope %s is broader than /%d. Narrow it to the segment you "
                    "are authorized to assess." % (piece, MAX_PREFIX))
            first = ip_to_int(info["network"])
            last = ip_to_int(info["broadcast"])
            if info["prefix"] < 31:
                first += 1
                last -= 1

        if last < first:
            raise ScopeError("Scope %s contains no usable address." % piece)

        for value in range(first, last + 1):
            ip = int_to_ip(value)
            if not allow_public and safe_range_of(ip) is None:
                raise ScopeError(
                    "Address %s is outside private address space. Start the "
                    "service with --allow-public to assess routable addresses, "
                    "and only with written authorization for those systems." % ip)
            if ip in seen:
                continue
            seen.add(ip)
            addresses.append(ip)
            if len(addresses) > MAX_ADDRESSES:
                raise ScopeError("Scope exceeds %d addresses." % MAX_ADDRESSES)

        entries.append({
            "input": piece, "info": info,
            "first": int_to_ip(first), "last": int_to_ip(last),
            "count": last - first + 1,
        })

    return Scope(", ".join(e["input"] for e in entries), entries, addresses)
